service-statistics-unsd/src/main/java/fi/nls/oskari/control/statistics/plugins/unsd/UnsdIndicatorValuesFetcher.java

```diff
@@ -0,0 +1,81 @@
+package fi.nls.oskari.control.statistics.plugins.unsd;
+
+import fi.nls.oskari.control.statistics.data.IndicatorValue;
+import fi.nls.oskari.control.statistics.data.IndicatorValueFloat;
+import fi.nls.oskari.control.statistics.data.StatisticalIndicatorDataModel;
+import fi.nls.oskari.control.statistics.plugins.APIException;
+import fi.nls.oskari.control.statistics.plugins.unsd.parser.UnsdParser;
+import fi.nls.oskari.control.statistics.plugins.unsd.requests.UnsdRequest;
+import fi.nls.oskari.service.ServiceRuntimeException;
+import fi.nls.oskari.util.JSONHelper;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+public class UnsdIndicatorValuesFetcher {
+    private UnsdConfig config;
+
+    public void init(UnsdConfig config) {
+        this.config = config;
+    }
+
+    /**
+     * @param selectors        Used to query UNSD with.
+     * @param indicator        The indicator we want.
+     * @param areaCodes        List of areas we are interested in
+     * @return
+     */
+    public Map<String, IndicatorValue> get(StatisticalIndicatorDataModel selectors, String indicator, String[] areaCodes) {
+        UnsdRequest request = new UnsdRequest(config);
+        request.setIndicator(indicator);
+        request.setAreaCodes(areaCodes);
+
+        Map<String, IndicatorValue> result = new HashMap<>();
+        boolean allFetched = false;
+        String response;
+        while(!allFetched) {
+            try {
+                response = request.getIndicatorData(selectors);
+                result.putAll(parseIndicatorData(response));
+                allFetched = isLastPage(response);
+                if (!allFetched) {
+                    request.nextPage();
+                    response = request.getIndicatorData(selectors);
```

Comment:
`request.getIndicatorData(selectors)` is called twice per loop?

Reply:
Few suggestions to improve readability: 1) Remove String response; declaration from outside the loop. 2) Remove allFetched flag, instead use plain while (true) and break; if (isLastPage(response)) 3) try-catch the whole loop, not each iteration - on exception we break the loop anyway